Add type annotations for Function.filter

mjpieters opened the pull request:
This includes all methods / properties on the returned FunctionFilter
object.

This contributes towards #6810

This pull request is:

- [x] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

mjpieters committed:
Add type annotations for Function.filter

This includes all methods / properties on the returned FunctionFilter
object.

zzzeek commented:
looks good

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 91e391879b5227464179326195763418b4d9e309 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 91e391879b5227464179326195763418b4d9e309: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 1d4c30795fdd72b02c6ba8138a5f32286c3434f8 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 1d4c30795fdd72b02c6ba8138a5f32286c3434f8 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

mjpieters commented:
Hrm, so the pep484 linter (mypy) points out a possible issue in the _implementation_ of `FunctionFilter.self_group()`, because `against` is optional (default `None`) and `is_precedent()` is not meant to accept `None` values for the `against` argument.

Given that several other places pass in `against=None` defaults means that either the annotations for `is_precedent()` needs updating to make `against` accept `None` (`against: Optional[OperatorType]`) or `against` should not be optional in all `self_group()` implementations.

mjpieters committed:
update is_precedent() annotation to accept against=None

This is commonly passed in already by `.self_group()` calls where
`against` defaults to `None`.

Handle `against is None` explicitly as the standard type stubs only
accepts the key type for the second argument to `dict.get()`.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 6137b7b995b6ea0bd4e4195c5693d2312fa26639 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 6137b7b995b6ea0bd4e4195c5693d2312fa26639 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester commented on lib/sqlalchemy/sql/elements.py:
```diff
@@ -4472,15 +4472,21 @@ def over(
+    if not TYPE_CHECKING:
+        # A001 is a false-positive here, type is a class member
+        @util.memoized_property
+        def type(self) -> TypeEngine[_T]:  # noqa: A001
```
**Federico Caselli** (CaselIT) wrote:

strange that this is needed, what issue does the type checker has?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester replied:
**Federico Caselli** (CaselIT) wrote:

by "this" I mean the `if not Type_ckeching` block

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester replied:
**Michael Bayer** (zzzeek) wrote:

probably needs a different memoized decorator, ro

edit: this is also on the superclass.  i will try to fix it as this is a much newer mypy

edit: no we definitely need this, many many subclasses have a simple "type: TypeEngine" variable and simple assignment is used.   mypy does not understand the `@memoized_property` thing so we have to conceal it here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester replied:
**Michael Bayer** (zzzeek) wrote:

we don't actually need the type-checked annotation here though, as the base class has this already

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester commented on lib/sqlalchemy/sql/functions.py:
```diff
@@ -449,7 +452,21 @@ def within_group(self, *order_by):
+    def filter(
+        self,
+        __criterion0: _ColumnExpressionArgument[bool],
+        *criterion: _ColumnExpressionArgument[bool],
```
**Federico Caselli** (CaselIT) wrote:

if this is 2.1 one we could use `/` to avoid the __ convention

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester replied:
**Michael Bayer** (zzzeek) wrote:

let's do it in a subsequent commit, want to keep a clean cherry pick for the moment

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

Great work, thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

hey there -

we still are in a bad place as most of functions.py is not annotated and we are still getting complaints: https://github.com/sqlalchemy/sqlalchemy/discussions/10694

functions.py is a straightforward series of annotations is it possible we can shoot to get the whole module done ?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

this can be rebased off of https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4957 after that gerrit merges

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4972 has been **merged**. Congratulations! :)

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5001 has been **merged**. Congratulations! :)
